src/user/tag.ts

```diff
@@ -0,0 +1,182 @@
+   *
+   * Get Tag by id
+   * > Tips:
+   * This function is depending on the Puppet Implementation, see [puppet-compatible-table](https://github.com/Chatie/wechaty/wiki/Puppet#3-puppet-compatible-table)
+   *
+   * @static
+   * @param {string} id
+   * @returns {Tag}
+   * @example
+   * const bot = new Wechaty()
+   * await bot.start()
+   * const tag = bot.Tag.load('tagId')
+   */
+  public static load<T extends typeof Tag> (
+    this : T,
+    id   : string,
+  ): T['prototype'] {
+    if (!this.pool) {
+      log.verbose('Tag', 'load(%s) init pool', id)
+      this.pool = new Map<string, Tag>()
+    }
+    if (this === Tag) {
+      throw new Error(
+        'The global Tag class can not be used directly!'
+        + 'See: https://github.com/Chatie/wechaty/issues/1217',
+      )
+    }
+    if (this.pool === Tag.pool) {
+      throw new Error('the current pool is equal to the global pool error!')
+    }
+    const existingTag = this.pool.get(id)
+    if (existingTag) {
+      return existingTag
+    }
+
+    const newTag = new (this as any)(id) as Tag
+    this.pool.set(id, newTag)
+
+    return newTag
+  }
+
+  /**
+   * Get all tags of this account
+   *
+   * > Tips:
+   * This function is depending on the Puppet Implementation, see [puppet-compatible-table](https://github.com/Chatie/wechaty/wiki/Puppet#3-puppet-compatible-table)
+   *
+   * @static
+   * @returns {Promise<Tag[]>}
+   * @example
+   * const bot = new Wechaty()
+   * await bot.start()
+   * await bot.Tag.findAll()
+   */
+  public static async findAll<T extends typeof Tag> (
+    this: T,
+    target: typeof Contact | typeof Favorite,
+  ): Promise<Array<T['prototype']>> {
+    log.verbose('Tag', 'findAll(%s)', target)
+
+    const tagList: Tag[] = []
+
+    if (target === Contact || target === this.wechaty.Contact) {
+      const contactList = await this.wechaty.Contact.findAll()
+      for (const contact of contactList) {
+        const contactTagList = await contact.tags()
+        contactTagList.forEach(tag => {
+          if (!tagList.includes(tag)) {
+              tagList.push(tag)
+            }
+          })
+      }
+    // } else if (target === Favorite || target === this.wechaty.Favorite)
+    } else {
+      log.error('Tag', 'findAll() target is neither Contact nor Favorite')
+    }
+
+    return tagList
+  }
+
+  /**
+   * Get a new tag
+   *
+   * > Tips:
+   * This function is depending on the Puppet Implementation, see [puppet-compatible-table](https://github.com/Chatie/wechaty/wiki/Puppet#3-puppet-compatible-table)
+   *
+   * @static
+   * @param {string} [tag] the tag name which want to create
+   * @returns {Promise<Tag>}
+   * @example
+   * const bot = new Wechaty()
+   * await bot.Tag.get(tag)
+   */
+  public static async get<T extends typeof Tag> (
+    this: T,
+    tag: string,
+  ): Promise<T['prototype']> {
+    log.verbose('Tag', 'get(%s)', tag)
+    return this.load(tag)
```

Comment:
In the `get` function, there is no actual puppet call, then when shall we create the tag?

Reply:
In my current design, I want to use Lazy Operation for the Tag.

in the `get()` we only save the tag string in the class. The actual operation will happen when we call `add()` or `del()`, and we pass the tag string to the puppet API, then puppet API deal with the tag name with the Contact or Favorite.

I think this will be a clear design with the high-level API, please feel free to let me know what you think.

Reply:
Lazy Operation is perfect.

So we could use this method like this way : 
```js
const tag = await bot.Tag.get('myTag')
const friend =bot.Contact.load(xxx)
await tag.add(friend) // we will create a new tag named myTag in this method
```

So we need to check the tag exist or not each time when we use `add()`?

I have a question about it, how could we know the tag exist or not ? Or just create new tag each time when we use `add()`?

This design will make `tagAddContact()` a little heavy, it means `create()` + `add()`, if something wrong happened in `create()`, it will wrong many times.

BTW, this method should check the tag's length, if user input a string which length over 15, we can not give a right tag back to user. It will be exposed when they use `add()` method.